### PR TITLE
executor: Refactor probe channel & fix bug in chunks of join

### LINKED ISSUE #17849: executor: update the usage of the channel to improve the produce-consume relationship among operators
## Description
As explained in the source code reading of [hash join](https://pingcap.com/blog-cn/tidb-source-code-reading-9/) and [hash aggregation](https://pingcap.com/blog-cn/tidb-source-code-reading-22/), the relationship is
1 input : N worker : 1 output. Before one work joining tuples, the worker has to ask one empty chunk from the output, and get one non-empty chunk from the input. Note that it has to synchronously wait for the input or the output once the latters are not ready. 

Besides, We observe 20ms ~ 30ms free time for works by tracing the join. @SunRunAway 

To improve this produce-consume relationship, we may do as followers:

- [ ] increase the channel resouces among them. This may observe some improvements, but it is not better to do the following way. 
- [ ] update the usage of `go channel`, among 1 input and N worker, it just need one channle, instead of 1 + N channels, which is the current implementation. This causes more resources, complexes the code, makes it hard to understand. So we should change the current implementation by using one channel. Similar problem exists among N work : 1 outpe.  Then we dynamically set the size of the channel among producers and consumers to improve the online throughput.

## Score 

* score number 1000

## Mentor(s)

* @fzhedu 

Contact the mentors: **#tidb-challenge-program** channel in [TiDB Community](https://join.slack.com/t/tidbcommunity/shared_invite/enQtNzc0MzI4ODExMDc4LWYwYmIzMjZkYzJiNDUxMmZlN2FiMGJkZjAyMzQ5NGU0NGY0NzI3NTYwMjAyNGQ1N2I2ZjAxNzc1OGUwYWM0NzE)  Slack Workspace

## Recommended Skills

* skills 1 know the usage of `go channel`
* skills 1 understand the produce-consume relationship in hash Join and hash aggregation.

## Learning Materials

* Chinese: [TiDB 精选技术讲解文章](https://github.com/pingcap/presentations/blob/master/hackathon-2019/reference-document-of-hackathon-2019.md)
* English: [Awesome-Database-Learning](https://github.com/pingcap/awesome-database-learning)

### PR DESCRIPTION
**Pull request by jianzhiyao**
### What problem does this PR solve?

Issue Number:   close https://github.com/pingcap/tidb/issues/17849

### What is changed and how it works?

Proposal: [executor: update the usage of the channel to improve the produce-consume relationship among operators](https://github.com/pingcap/tidb/issues/17849) 

What's Changed:

- change ``HashJoinExec.probeResultChs`` and ``joinChkResourceCh`` from multi channel to single channel
- fix scope ambiguity of ``joinResult`` in ``joinMatchedProbeSideRow2ChunkForOuterHashJoin`` and ``joinMatchedProbeSideRow2Chunk``

### Check List

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

```
name                                                                                                         old time/op    new time/op    delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12     1.29s ± 1%     1.32s ± 3%  +2.28%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12       268ms ± 1%     277ms ± 0%  +3.29%  (p=0.000 n=8+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        6.41s ± 3%     6.24s ± 6%  -2.59%  (p=0.043 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  63.5µs ± 5%    60.9µs ± 4%  -4.11%  (p=0.006 n=10+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12           36.6ms ± 3%    36.3ms ± 1%    ~     (p=0.165 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12             32.6ms ± 2%    32.3ms ± 1%    ~     (p=0.546 n=9+9)

name                                                                                                         old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12    2.46GB ± 0%    2.46GB ± 0%    ~     (p=0.684 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12       307MB ± 0%     307MB ± 0%    ~     (p=0.579 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        903MB ± 0%     903MB ± 0%    ~     (p=0.684 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  80.1kB ± 0%    79.3kB ± 0%  -0.99%  (p=0.000 n=10+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12           19.6MB ± 0%    19.6MB ± 0%    ~     (p=0.912 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12             17.9MB ± 0%    17.9MB ± 0%    ~     (p=0.684 n=10+10)

name                                                                                                         old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12      813k ± 0%      813k ± 0%  -0.01%  (p=0.033 n=10+9)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12        413k ± 0%      413k ± 0%  -0.01%  (p=0.002 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        2.01M ± 0%     2.01M ± 0%    ~     (p=0.343 n=10+10)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                     908 ± 0%       886 ± 0%  -2.42%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12             614k ± 0%      613k ± 0%  -0.00%  (p=0.000 n=10+10)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12               413k ± 0%      413k ± 0%  -0.01%  (p=0.000 n=10+10)
```

Side effects

- Performance regression
    - Consumes less CPU
    - Consumes less MEM

### Release note 

- update the usage of the channel in hash join to improve the produce-consume relationship among operators
- fix scope ambiguity of ``joinResult`` in ``joinMatchedProbeSideRow2ChunkForOuterHashJoin`` and ``joinMatchedProbeSideRow2Chunk``

### CODE REVIEW

**Comment by jianzhiyao** [comment]
PTAL @fzhedu @lzmhhh123

**Comment by jianzhiyao** [comment]
After testing, i find there is no need to change mode of channel from 1:n to 1:1.
Maybe there are no enhancement on changing mode of channel.

### Testing code
```
func worker(wg *sync.WaitGroup, done chan bool, ch <-chan bool) {
	for ok := true; ok; {
		select {
		case <-done:
			ok = false
		case <-ch:
			wg.Done()
		}
	}
	wg.Done()
}

func BenchmarkMultiChannel(b *testing.B){
	var wg sync.WaitGroup
	concurrency := 5
	done := make(chan bool)
	chs := make([]chan bool, concurrency)
	for i := 0; i < concurrency; i++ {
		chs[i] = make(chan bool, 1)
		go worker(&wg, done, chs[i])
	}

	wg.Add(b.N * concurrency)
	for i := 0; i < b.N; i++ {
		for j := 0; j < concurrency; j++ {
			go func(j int) {
				chs[j] <- true
			}(j)
		}
	}
	wg.Wait()

	wg.Add(concurrency)
	close(done)
	wg.Wait()

	for i:= range chs {
		close(chs[i])
		for range chs[i] {
		}
	}
}
```

### benchmark output
```
BenchmarkMultiChannel
BenchmarkMultiChannel-6   	  545498	      2279 ns/op
BenchmarkMultiChannel-6   	  521736	      2448 ns/op
BenchmarkMultiChannel-6   	  545468	      2427 ns/op
BenchmarkMultiChannel-6   	  522376	      2313 ns/op
BenchmarkMultiChannel-6   	  571411	      2548 ns/op
BenchmarkMultiChannel-6   	  571408	      2247 ns/op
BenchmarkMultiChannel-6   	  571414	      2350 ns/op
BenchmarkMultiChannel-6   	  571400	      2473 ns/op
BenchmarkMultiChannel-6   	  444430	      2457 ns/op
BenchmarkMultiChannel-6   	  599960	      2514 ns/op
-------
BenchmarkSingleChannel
BenchmarkSingleChannel-6   	  444444	      3085 ns/op
BenchmarkSingleChannel-6   	  399979	      3470 ns/op
BenchmarkSingleChannel-6   	  387091	      3485 ns/op
BenchmarkSingleChannel-6   	  363639	      3407 ns/op
BenchmarkSingleChannel-6   	  363640	      3413 ns/op
BenchmarkSingleChannel-6   	  374991	      3309 ns/op
BenchmarkSingleChannel-6   	  342853	      3459 ns/op
BenchmarkSingleChannel-6   	  375038	      3394 ns/op
BenchmarkSingleChannel-6   	  400004	      3507 ns/op
BenchmarkSingleChannel-6   	  352953	      3785 ns/op

```

**Comment by fzhedu** [comment]
It is nice to bench multiple channels vs. single channel. But we should clarify the work cases:
1. an input goroutine generates `w` tasks, and each task takes several secondes, like task_i{i%3 sec};
2. n workers fetch `w ` tasks and each sleeps {i%3 sec};
3. an output goroutine fetch work results from 2.

we can bench the case from following aspects:
1. w >> n , and n increase from 1 to 1000 or even higher
2. fluctuant input and output speed.
3. channel size, at lease n

File: executor/join.go [hunk 1]
```diff
@@ -73,9 +73,9 @@ type HashJoinExec struct {
 	// execution, to avoid the concurrency of joiner.chk and joiner.selected.
 	joiners []joiner
 
-	probeChkResourceCh chan *probeChkResource
-	probeResultChs     []chan *chunk.Chunk
-	joinChkResourceCh  []chan *chunk.Chunk
+	probeChkResourceCh chan *chunk.Chunk
+	probeChkResultCh   chan *chunk.Chunk
+	joinChkResourceCh  chan *chunk.Chunk
```
Comment (fzhedu) [inline executor/join.go:78]: can you reuse the chunks like https://github.com/pingcap/tidb/pull/19178/ ?

Comment (jianzhiyao) [reply executor/join.go:78]: Seems chunk is already reused.

**Comment by jianzhiyao** [comment]
> It is nice to bench multiple channels vs. single channel. But we should clarify the work cases:
> 
> 1. an input goroutine generates `w` tasks, and each task takes several secondes, like task_i{i%3 sec};
> 2. n workers fetch `w ` tasks and each sleeps {i%3 sec};
> 3. an output goroutine fetch work results from 2.
> 
> we can bench the case from following aspects:
> 
> 1. w >> n , and n increase from 1 to 1000 or even higher
> 2. fluctuant input and output speed.
> 3. channel size, at lease n

After considering some factor in fact.
## There is the code.
```
func BenchmarkChannelMode(b *testing.B) {
	currencies := [...]int{1, 5, 10, 100, 1000}
	workerChanSizes := [...]int{1, 5, 10}
	for _, currency := range currencies {
		for _, size := range workerChanSizes {
			b.Run(fmt.Sprintf("benchmarkChannel[type=multi][con-%d][size-%d]", currency, size), func(b *testing.B) {
				benchmarkMultiChannel(b, currency, size)
			})
			b.Run(fmt.Sprintf("benchmarkChannel[type=single][con-%d][size-%d]", currency, size), func(b *testing.B) {
				benchmarkSingleChannel(b, currency, size)
			})
		}
	}
}

func probeWorker(wg *sync.WaitGroup, done <-chan bool, probeCh <-chan bool, resultCh chan<- bool, workerId int) {
	for ok := true; ok; {
		select {
		case <-done:
			ok = false
		case <-probeCh:
			time.Sleep(time.Duration(workerId%3) * 1 * time.Millisecond)
			resultCh <- true
		}
	}
	wg.Done()
}

func resultWorker(wg *sync.WaitGroup, done <-chan bool, resultCh <-chan bool) {
	for ok := true; ok; {
		select {
		case <-done:
			ok = false
		case <-resultCh:
			wg.Done()
		}
	}
	wg.Done()
}

func benchmarkMultiChannel(b *testing.B, concurrency int, workerChanSize int) {
	var wg sync.WaitGroup
	done := make(chan bool)
	probeCh := make([]chan bool, concurrency)
	resultCh := make(chan bool, concurrency)

	go resultWorker(&wg, done, resultCh)
	for i := 0; i < concurrency; i++ {
		probeCh[i] = make(chan bool, workerChanSize)
		go probeWorker(&wg, done, probeCh[i], resultCh, i)
	}

	wg.Add(b.N * concurrency)
	for i := 0; i < b.N; i++ {
		for j := 0; j < concurrency; j++ {
			go func(j int) {
				time.Sleep(time.Duration(j%3) * 1 * time.Millisecond)
				probeCh[j] <- true
			}(j)
		}
	}
	wg.Wait()

	wg.Add(concurrency + 1) //probeWorker + resultWorker
	close(done)
	wg.Wait()

	close(resultCh)
	for range resultCh {
	}
	for i := range probeCh {
		close(probeCh[i])
		for range probeCh[i] {
		}
	}
}

func benchmarkSingleChannel(b *testing.B, concurrency int, workerChanSize int) {
	var wg sync.WaitGroup
	done := make(chan bool)
	probe := make(chan bool, concurrency*workerChanSize)
	res := make(chan bool, concurrency)

	go resultWorker(&wg, done, res)
	for i := 0; i < concurrency; i++ {
		go probeWorker(&wg, done, probe, res, i)
	}

	wg.Add(b.N * concurrency)
	for i := 0; i < b.N; i++ {
		for j := 0; j < concurrency; j++ {
			go func(j int) {
				time.Sleep(time.Duration(j%3) * 100 * time.Millisecond)
				probe <- true
			}(j)
		}
	}
	wg.Wait()

	wg.Add(concurrency + 1) //probeWorker + resultWorker
	close(done)
	wg.Wait()

	close(res)
	for range res {
	}
	close(probe)
	for range probe {
	}
}
```
## benchmark output
Seems single channel is better, after considering real operating action.
```
BenchmarkChannelMode
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-1]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-1]-12         	  922920	      1284 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-1]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-1]-12        	 1000000	      1238 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-5]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-5]-12         	 1000000	      1235 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-5]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-5]-12        	 1000000	      1318 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-10]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1][size-10]-12        	  857062	      1390 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-10]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1][size-10]-12       	  923104	      1272 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-1]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-1]-12         	     459	   2554742 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-1]
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-1]-12        	  193222	      6469 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-5]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-5]-12         	     445	   2612823 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-5]
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-5]-12        	  192426	      6460 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-10]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-5][size-10]-12        	     489	   2551720 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-10]
BenchmarkChannelMode/benchmarkChannel[type=single][con-5][size-10]-12       	  192427	      6460 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-1]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-1]-12        	     454	   2567317 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-1]
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-1]-12       	   98341	     12299 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-5]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-5]-12        	     468	   2569419 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-5]
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-5]-12       	   99151	     12214 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-10]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-10][size-10]-12       	     448	   2563199 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-10]
BenchmarkChannelMode/benchmarkChannel[type=single][con-10][size-10]-12      	   98839	     12258 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-1]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-1]-12       	     481	   2533874 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-1]
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-1]-12      	   10323	    130197 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-5]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-5]-12       	     471	   2559946 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-5]
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-5]-12      	   10498	    129502 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-10]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-100][size-10]-12      	     453	   2575389 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-10]
BenchmarkChannelMode/benchmarkChannel[type=single][con-100][size-10]-12     	   10386	    125845 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-1]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-1]-12      	     441	   2669683 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-1]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-1]-12     	     900	   1363900 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-5]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-5]-12      	     427	   2714862 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-5]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-5]-12     	     793	   1546704 ns/op
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-10]
BenchmarkChannelMode/benchmarkChannel[type=multi][con-1000][size-10]-12     	     434	   2670222 ns/op
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-10]
BenchmarkChannelMode/benchmarkChannel[type=single][con-1000][size-10]-12    	     957	   1386068 ns/op
PASS
```

**Comment by ichn-hu** [comment]
@jianzhiyao what's your conclusion after the new benchmark?

**Comment by jianzhiyao** [comment]
> @jianzhiyao what's your conclusion after the new benchmark?

###  Lastest benchmark
Seens changing channel mode is no a good practice for hash join. 

So I fix scope ambiguity of joinResult in https://github.com/pingcap/tidb/pull/23876. And I think https://github.com/pingcap/tidb/issues/17849 can be closed After https://github.com/pingcap/tidb/pull/23876 being merged.

PTAL @ichn-hu
```
name                                                                                                         old time/op    new time/op    delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12     1.29s ± 2%     1.33s ± 7%  +2.97%  (p=0.000 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12       265ms ± 1%     289ms ± 3%  +9.19%  (p=0.000 n=19+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        6.34s ± 2%     6.51s ± 4%  +2.69%  (p=0.000 n=20+20)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  62.7µs ± 5%    64.4µs ± 5%  +2.78%  (p=0.012 n=20+19)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12           38.9ms ± 2%    37.7ms ± 3%  -3.08%  (p=0.000 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12             34.4ms ± 1%    33.9ms ± 2%  -1.39%  (p=0.000 n=20+20)

name                                                                                                         old alloc/op   new alloc/op   delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12    2.46GB ± 0%    2.46GB ± 0%    ~     (p=0.322 n=20+19)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12       307MB ± 0%     307MB ± 0%  -0.00%  (p=0.048 n=20+18)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        903MB ± 0%     903MB ± 0%    ~     (p=0.698 n=20+20)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  80.1kB ± 0%    79.3kB ± 0%  -0.98%  (p=0.000 n=20+19)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12           19.6MB ± 0%    19.6MB ± 0%    ~     (p=0.841 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12             17.9MB ± 0%    17.9MB ± 0%    ~     (p=0.149 n=20+18)

name                                                                                                         old allocs/op  new allocs/op  delta
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12      813k ± 0%      813k ± 0%  -0.01%  (p=0.007 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12        413k ± 0%      413k ± 0%  -0.00%  (p=0.018 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12        2.01M ± 0%     2.01M ± 0%  -0.00%  (p=0.015 n=20+20)
HashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                     908 ± 0%       886 ± 0%  -2.42%  (p=0.000 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12             614k ± 0%      613k ± 0%  -0.00%  (p=0.000 n=20+20)
HashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12               413k ± 0%      413k ± 0%  -0.00%  (p=0.000 n=20+19)
```

**Pull request closed without merging**